### PR TITLE
Feature/orb 64 agent group creation endpoint

### PR DESCRIPTION
**Pull request by mclcavalcante**
feat(agent_group): Agent group validation

### CODE REVIEW

File: fleet/agent_group_service_test.go [hunk 1]
```diff
@@ -116,3 +122,51 @@ func TestCreateAgentGroup(t *testing.T) {
...
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s", err))
+	nameID, err := types.NewIdentifier("eu-agents")
+	require.Nil(t, err, fmt.Sprintf("unexpected error: %s", err))
+
+	validAgent := fleet.AgentGroup{
+		MFOwnerID:   ownerID.String(),
+		Name:        nameID,
+		Description: "An example agent group representing european dns nodes",
+		Tags:        make(map[string]string),
+		Created:     time.Time{},
```
Comment (dscabral) [inline fleet/agent_group_service_test.go:142]: For validate, this Created it's not necessary

File: fleet/agent_service.go [hunk 2]
```diff
@@ -98,3 +98,13 @@ func (svc fleetService) CreateAgent(ctx context.Context, token string, a Agent)
...
+
+func (svc fleetService) ValidateAgent(ctx context.Context, token string, a Agent) (Agent, error) {
+	mfOwnerID, err := svc.identify(token)
+	if err != nil {
+		return Agent{}, err
+	}
+
+	a.MFOwnerID = mfOwnerID
+	return a, nil
+}
```
Comment (dscabral) [inline fleet/agent_service.go:110]: Should remove, it's not part of this issue.

File: fleet/api/endpoint.go [hunk 3]
```diff
@@ -178,3 +178,35 @@ func listAgentsEndpoint(svc fleet.Service) endpoint.Endpoint {
...
+		saved, err := svc.ValidateAgentGroup(c, req.token, group)
+		if err != nil {
+			return nil, err
+		}
+
+		res := validateAgentGroupRes{
+			ID:      saved.ID,
+			Name:    saved.Name.String(),
+			Tags:    saved.Tags,
+			created: true,
```
Comment (dscabral) [inline fleet/api/endpoint.go:207]: Should not return created, if we're just validating.

File: fleet/api/logging.go [hunk 4]
```diff
@@ -106,6 +106,38 @@ func (l loggingMiddleware) CreateAgentGroup(ctx context.Context, token string, s
 	return l.svc.CreateAgentGroup(ctx, token, s)
 }
 
+func (l loggingMiddleware) ValidateAgent(ctx context.Context, token string, a fleet.Agent) (_ fleet.Agent, err error) {
```
Comment (dscabral) [inline fleet/api/logging.go:109]: Remove this validate agent

File: fleet/api/metrics.go [hunk 5]
```diff
@@ -42,6 +42,14 @@ func (m metricsMiddleware) CreateAgentGroup(ctx context.Context, token string, s
 	return m.svc.CreateAgentGroup(ctx, token, s)
 }
 
+func (m metricsMiddleware) ValidateAgent(ctx context.Context, token string, a fleet.Agent) (fleet.Agent, error) {
```
Comment (dscabral) [inline fleet/api/metrics.go:45]: Remove this validate agent

File: fleet/api/openapi.yaml [hunk 6]
```diff
@@ -103,6 +103,32 @@ paths:
...
+      requestBody:
+        $ref: "#/components/requestBodies/AgentGroupCreateReq"
+      responses:
+        '200':
+          description: Valid Agent group
+        '400':
+          description: Failed due to malformed JSON.
+        '401':
+          description: Missing or invalid access token provided.
+        '409':
```
Comment (dscabral) [inline fleet/api/openapi.yaml:123]: Remove this error, it will not conflict.

File: fleet/api/openapi.yaml [hunk 7]
```diff
@@ -103,6 +103,32 @@ paths:
...
+          description: Entity already exist.
+        '415':
+          description: Missing or invalid content type.
+        '422':
```
Comment (dscabral) [inline fleet/api/openapi.yaml:127]: Remove this error, it will not make a query on db.

File: fleet/api/responses.go [hunk 8]
```diff
@@ -128,3 +128,50 @@ type pageRes struct {
...
 }
+
+type validateAgentGroupRes struct {
+	ID             string         `json:"id"`
+	Name           string         `json:"name"`
+	Description    string         `json:"description,omitempty"`
+	Tags           types.Tags     `json:"tags"`
+	TsCreated      time.Time      `json:"ts_created,omitempty"`
+	MatchingAgents types.Metadata `json:"matching_agents,omitempty"`
+	created        bool
```
Comment (dscabral) [inline fleet/api/responses.go:139]: Don't need created and TsCreated fields for this struct

File: fleet/api/responses.go [hunk 9]
```diff
@@ -128,3 +128,50 @@ type pageRes struct {
...
+
+func (s validateAgentGroupRes) Headers() map[string]string {
+	return map[string]string{}
+}
+
+func (s validateAgentGroupRes) Empty() bool {
+	return false
+}
+
+type validateAgentRes struct {
```
Comment (dscabral) [inline fleet/api/responses.go:154]: Remove this struct

File: fleet/mocks/agent_groups.go [hunk 10]
```diff
@@ -71,3 +78,13 @@ func (a *agentGroupRepositoryMock) RetrieveAllAgentGroupsByOwner(ctx context.Con
 	}
 	return pageAgentGroup, nil
 }
+
```
Comment (dscabral) [inline fleet/mocks/agent_groups.go:81]: Its not necessary any more

File: fleet/redis/producer/streams.go [hunk 11]
```diff
@@ -50,6 +50,14 @@ func (es eventStore) CreateAgentGroup(ctx context.Context, token string, s fleet
 	return es.svc.CreateAgentGroup(ctx, token, s)
 }
 
+func (es eventStore) ValidateAgent(ctx context.Context, token string, a fleet.Agent) (fleet.Agent, error) {
```
Comment (dscabral) [inline fleet/redis/producer/streams.go:53]: Remove this function

**Review by dscabral — CHANGES_REQUESTED** [review]

File: fleet/api/responses.go [hunk 12]
```diff
@@ -128,3 +128,24 @@ type pageRes struct {
 	Order  string `json:"order"`
 	Dir    string `json:"direction"`
 }
+
+type validateAgentGroupRes struct {
+	ID             string         `json:"id"`
+	Name           string         `json:"name"`
+	Description    string         `json:"description,omitempty"`
+	Tags           types.Tags     `json:"tags"`
+	TsCreated      time.Time      `json:"ts_created,omitempty"`
```
Comment (dscabral) [inline fleet/api/responses.go:137]: We don't need this TsCreated on validateRes

**Review by dscabral — CHANGES_REQUESTED** [review]

**Review by dscabral — APPROVED** [review]

**Review by dscabral — APPROVED** [review]

**Review by dscabral — DISMISSED** [review]

File: fleet/agent_group_service.go [hunk 13]
```diff
@@ -182,3 +182,14 @@ func (svc fleetService) RemoveAgentGroup(ctx context.Context, token, groupId str
 
 	return nil
 }
+
+func (svc fleetService) ValidateAgentGroup(ctx context.Context, token string, s AgentGroup) (AgentGroup, error) {
+	mfOwnerID, err := svc.identify(token)
+	if err != nil {
+		return AgentGroup{}, err
+	}
+
```
Comment (weyrick) [inline fleet/agent_group_service.go:191]: I thought the goal here was also to return the number of matching agents?

File: fleet/postgres/agents.go [hunk 14]
```diff
@@ -30,6 +30,41 @@ type agentRepository struct {
 	logger *zap.Logger
 }
 
+func (r agentRepository) RetrieveMatchingAgents(ctx context.Context, tags types.Tags) (types.Metadata, error) {
+	t, tmq, err := getOrbTagsQuery(tags)
```
Comment (weyrick) [inline fleet/postgres/agents.go:34]: To be clear, whenever we are matching an agent group to agents, it must match against _both_ `agents.agent_tags` and `agents.orb_tags` - is that what we are doing here and in the other calls? You will see that the query for the `agent_group_membership` does this, for example.

File: fleet/postgres/agents.go [hunk 15]
```diff
@@ -30,6 +30,41 @@ type agentRepository struct {
...
+	}
+
+	q := fmt.Sprintf(
+		`select
+			json_build_object('total', coalesce(total,0), 'online', coalesce(online,0)) AS matching_agents
+		from
+			(select
+				sum(case when mf_thing_id is not null then 1 else 0 end) as total,
+				sum(case when state = 'online' then 1 else 0 end) as online
+			from agents WHERE mf_thing_id is not null %s) as agent_groups`, tmq)
```
Comment (weyrick) [inline fleet/postgres/agents.go:46]: security problem: this query does not limit for OwnerID so it will query agents from all tenants

**Review by weyrick — CHANGES_REQUESTED** [review]

**Review by weyrick — APPROVED** [review]

**Review by weyrick — APPROVED** [review]

**Review by dscabral — APPROVED** [review]

**Pull request merged**